### PR TITLE
Clarity: Unsigned Integers

### PR DESCRIPTION
**Pull request by kantai**
This PR adds support for a new Clarity type for unsigned integers.

In implementation of it, the following changes were made:

1. Added a parser literal for unsigned ints: `u9011..` -> unsigned integer.
2. Updated the arithmetic operations to take _either_ unsigned ints or ints. Attempting to use both in the same op will throw a TypeError.
3. Updated the type checker to enforce the above.

There's no additional functions for unsigned integer casting, though they may be useful, i.e.:

```
(as-int u8) -> 8
(as-uint 8) -> u8
(as-uint -9) -> abort with Arithmetic runtime error
```

This PR _also_ applies some corrections to our typing system for response types, option types, and lists of those types, with tests for coverage of those cases -- in particular, the type system should allow "NoType"s to exist, since they are legal when, for example, the "some" type correlative of a none is unknown, or the analogous situation with response types.

This implements #974

### CODE REVIEW

**Comment by kantai** [comment]
Right -- I think such functions are definitely needed. There's two options, I think:

```
;; Option 1
(to-uint <int>) -> returns uint. causes runtime error (i.e., aborts transaction) in failure
(to-int <uint>) -> returns int. causes runtime error (i.e., aborts transaction) in failure

;; Option 2
(to-uint <int>) -> returns (option uint)
(to-int <uint>) -> returns (option int)
```

**Comment by kantai** [comment]
Implemented Option 1:

```
;; Option 1
(to-uint <int>) -> returns uint. causes runtime error (i.e., aborts transaction) in failure
(to-int <uint>) -> returns int. causes runtime error (i.e., aborts transaction) in failure
```

With tests for the type analysis of those functions as well as their runtime behaviors.

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -91,6 +91,20 @@ const NONE_KEYWORD: KeywordAPI = KeywordAPI {
...
+    name: None,
+    signature: "(to-uint i)",
+    description: "Tries to convert the `int` argument to a `uint`. Will cause a runtime error and abort if the supplied argument is negative.",
+    example: "(to-uint 238) ;; Returns u238"
+};
+
+const TO_INT_API: SimpleFunctionAPI = SimpleFunctionAPI {
+    name: None,
+    signature: "(to-int u)",
+    description: "Tries to convert the `uint` argument to an `int`. Will cause a runtime error and abort if the supplied argument is >= 2^127.",
```
Comment (jcnelson) [inline src/vm/docs/mod.rs]: Should be `2^127 - 1`.  Also, to me, `^` means `bitwise xor`.  Do you think `**` would make this less ambiguous, or more confusing?  Maybe `pow(2, 127)`?

Comment (moxiegirl) [reply src/vm/docs/mod.rs]: So, the choices are:

* `^`
* `**`
* `pow(int, int)`

If I am correct, I think the last is more readable. The last is actually similar to the Javascript language `Math.pow` another positive. @jcnelson @kantai

Comment (kantai) [reply src/vm/docs/mod.rs]: Great, that's what I thought as well

**Comment by kantai** [comment]
> Should we also migrate the implementations of our fungible tokens / non fungible tokens - from int to uint?

Yes -- probably. I'll start an issue for using uints in our intrinsics, but will merge this first.

**Pull request merged**